kernel/src/util/iovec.rs

```diff
@@ -37,92 +39,192 @@ impl TryFrom<UserIoVec> for IoVec {
+        Ok(Self(readers))
+    }
+
+    /// Returns mutable reference to [`VmReader`]s.
+    pub fn readers_mut(&'a mut self) -> &'a mut [VmReader<'a>] {
+        &mut self.0
+    }
+}
+
+impl<'a> VmWriterArray<'a> {
+    /// Creates a new `IoVecWriter` from user-provided io vec buffer.
+    pub fn from_user_io_vecs(
+        ctx: &'a Context<'a>,
+        start_addr: Vaddr,
+        count: usize,
+    ) -> Result<Self> {
+        let writers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::writer)?;
+        Ok(Self(writers))
+    }
+
+    /// Returns mutable reference to [`VmWriter`]s.
+    pub fn writers_mut(&'a mut self) -> &'a mut [VmWriter<'a>] {
+        &mut self.0
     }
+}
 
-    /// Reads bytes from the user space buffer pointed by
-    /// the `IoVec` to `dst`.
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait MultiRead {
+    /// Reads the exact number of bytes required to exhaust `self` or fill `writer`,
+    /// accumulating total bytes read.
     ///
-    /// If successful, the read length will be equal to `dst.len()`.
+    /// If the return value is `Ok(n)`,
+    /// then `n` should be `min(self.sum_lens(), writer.avail())`.
     ///
-    /// # Panics
+    /// # Errors
     ///
-    /// This method will panic if
-    /// 1.`dst.len()` is not the same as `self.len()`;
-    /// 2. `self.is_empty()` is `true`.
-    pub fn read_exact_from_user(&self, dst: &mut [u8]) -> Result<()> {
-        assert_eq!(dst.len(), self.len);
-        assert!(!self.is_empty());
+    /// This method returns [`Errno::EFAULT`] if a page fault occurs.
+    fn read(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
+
+    /// Calculates the total length of data remaining to read.
+    fn sum_lens(&self) -> usize;
 
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(dst))
+    /// Checks if the data remaining to read is empty.
+    fn is_empty(&self) -> bool {
+        self.sum_lens() == 0
     }
+}
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
+/// Trait defining the write behavior for a collection of [`VmWriter`]s.
+pub trait MultiWrite {
+    /// Writes the exact number of bytes required to exhaust `writer` or fill `self`,
+    /// accumulating total bytes read.
     ///
-    /// If successful, the written length will be equal to `src.len()`.
+    /// If the return value is `Ok(n)`,
+    /// then `n` should be `min(self.sum_lens(), reader.remain())`.
     ///
-    /// # Panics
+    /// # Errors
     ///
-    /// This method will panic if
-    /// 1. `src.len()` is not the same as `self.len()`;
-    /// 2. `self.is_empty()` is `true`.
-    pub fn write_exact_to_user(&self, src: &[u8]) -> Result<()> {
-        assert_eq!(src.len(), self.len);
-        assert!(!self.is_empty());
+    /// This method returns [`Errno::EFAULT`] if a page fault occurs.
```

Comment:
Documenting the state (position? cursor?) of the writers and the reader is left unspecified when the method returns `EFAULT`. 

At least for now, no one needs to know their position, so I think leaving it unspecified (and documenting the behavior) is fine. 

Otherwise, we may need to discuss whether we want to revert the position.

Reply:
The comment is refactored as suggested.